sdk/azcore/poller.go

```diff
@@ -0,0 +1,453 @@
+	URL() string
+	Status() string
+	Succeeded() bool
+}
+
+// ====================================================================================================
+
+// polls on the operation-location header
+type opPoller struct {
+	Type      string `json:"type"`
+	ReqMethod string `json:"reqMethod"`
+	ReqURL    string `json:"reqURL"`
+	PollURL   string `json:"pollURL"`
+	LocURL    string `json:"locURL"`
+	status    string
+}
+
+func newOpPoller(pollerType, pollingURL, locationURL string, initialResponse *Response) *opPoller {
+	return &opPoller{
+		Type:      fmt.Sprintf("%s;opPoller", pollerType),
+		ReqMethod: initialResponse.Request.Method,
+		ReqURL:    initialResponse.Request.URL.String(),
+		PollURL:   pollingURL,
+		LocURL:    locationURL,
+	}
+}
+
+func (p *opPoller) URL() string {
+	return p.PollURL
+}
+
+func (p *opPoller) Done() bool {
+	return strings.EqualFold(p.status, "succeeded") ||
+		strings.EqualFold(p.status, "failed") ||
+		strings.EqualFold(p.status, "cancelled")
+}
+
+func (p *opPoller) Succeeded() bool {
+	return strings.EqualFold(p.status, "succeeded")
+}
+
+func (p *opPoller) Update(resp *Response) error {
+	status, err := extractJSONValue(resp, "status")
+	if err != nil {
+		return err
+	}
+	if status == "" {
+		return errors.New("no status found in body")
+	}
+	p.status = status
+	// if the endpoint returned an operation-location header, update cached value
+	if opLoc := resp.Header.Get(headerOperationLocation); opLoc != "" {
+		p.PollURL = opLoc
+	}
+	return nil
+}
+
+func (p *opPoller) FinalGetURL(resp *Response) (string, error) {
+	if !p.Done() {
+		return "", errors.New("cannot return a final response from a poller in a non-terminal state")
+	}
+	resLoc, err := extractJSONValue(resp, "resourceLocation")
+	if err != nil {
+		return "", err
+	}
+	if resLoc != "" {
+		return resLoc, nil
+	}
+	if p.ReqMethod == http.MethodPatch || p.ReqMethod == http.MethodPut {
+		return p.ReqURL, nil
+	}
+	if p.ReqMethod == http.MethodPost && p.LocURL != "" {
+		return p.LocURL, nil
+	}
+	return "", nil
+}
+
+func (p *opPoller) Status() string {
+	return p.status
+}
+
+// ====================================================================================================
+
+// polls on the location header
+type locPoller struct {
+	Type    string `json:"type"`
+	PollURL string `json:"pollURL"`
+	status  int
+}
+
+func newLocPoller(pollerType, pollingURL string, initialStatus int) *locPoller {
+	return &locPoller{
+		Type:    fmt.Sprintf("%s;locPoller", pollerType),
+		PollURL: pollingURL,
+		status:  initialStatus,
+	}
+}
+
+func (p *locPoller) URL() string {
+	return p.PollURL
+}
+
+func (p *locPoller) Done() bool {
+	// a 202 means the operation is still in progress
+	// zero-value indicates the poller was rehydrated from a token
+	return p.status > 0 && p.status != http.StatusAccepted
+}
+
+func (p *locPoller) Succeeded() bool {
+	// any 2xx status code indicates success
+	return p.status >= 200 && p.status < 300
```

Comment:
This is interesting, we dont have this in the ARM pollers do we??

Reply:
I don't think ARM has such simple polling.  The logic for locPoller is that it will continue to poll on a 202 returned from the polling endpoint.  A non-202 response that's a 2xx indicates the operation succeeded.  All other status codes indicate it failed.

Reply:
Oh cool. Good to know, thanks!